### DISCUSSION TITLE
TS: Environment variables are undefined on building process

### DISCUSSION THREAD

**Discussion by abaracedo** [vercel/next.js#52250] [Help] [answered]
### Summary

I created a middleware.tsx to handle some routes. In its function I'm accessing an env var and works with `next dev` but when I run `next build` I get the error: `Type error: Type 'undefined' cannot be used as an index type.`

It seems that the issue is related to TS (I'm new on it) so I'd like to know which should be the proper way to solve it. I've seen that's possible to use `// @ts-ignore` but I don't know if I should avoid using that directime/comment.

**Comment by icyJoseph** [comment, marked as answer]
I guess you mean that `process.env.YOUR_ENV_VAR` is typed as `string | undefined`

Some choose to augment the ProcessEnv object, but that is often not a good idea. Instead you should do narrowing:

- https://stackoverflow.com/a/51144635

```ts
import { NextResponse } from "next/server";

export const middleware = () => {
  const foo = process.env.FOO; // string | undefined
  return NextResponse.next();
};

```

In the above, `foo` is a union of string and undefined, we need to narrow down to string.

```ts
import { NextResponse } from "next/server";

export const middleware = () => {
  const foo = process.env.FOO;
  
  if (typeof foo === 'undefined') {
    throw new Error("Env var `foo` is not defined")
  }

  foo // string
  return NextResponse.next();
};

```

Now after the `typeof` check, TS knows that `foo` is a string, so you can continue. Perhaps you don't want to throw if `foo` is missing, and instead redirect? or return an error? That's up to your implementation/requirement.

**Reply by abaracedo** [reply]
I see, type checking the value TS knows the value has to be a string. Interesting, I keep it for other uses I may need. Thanks for the help.

I'd like to think if the env var is defined, no matter how, it wouldn't evaluate to `undefined`.

**Reply by icyJoseph** [reply]
Right, but TS doesn't know that information, however, if you are sure, that you'll always have it, you could augment the `ProcessEnv` object as the link shows. Most people I think do a little type checking at top of a file, and if fails, you can't even start the project, until you provide the variable. 

For example here: https://github.com/vercel/next.js/blob/canary/examples/with-mongodb/lib/mongodb.ts
